### PR TITLE
Update typing_extensions

### PR DESCRIPTION
**Pull request by JelleZijlstra**

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Stubtest is unhappy because these names don't exist in the released version of typing-extensions yet. I think it's better to have them in typeshed already though, so that type checkers have more of a chance to pick up the new stubs in time for the next typing-extensions release.

Maybe we should install typing-extensions from git in the stubtest workflow.

**Comment by JelleZijlstra** [comment]
Seems like I need to fight stubtest some more, will look into it later today.

**Comment by JelleZijlstra** [comment]
Now it's only pytype that's complaining, but it seems to be dumping out the entire ast of `typing`: https://github.com/python/typeshed/actions/runs/5340691697/jobs/9680790454?pr=10344

@rchen152 could you take a look? This is trying to implement the recent typing-extensions change where we now re-export more of typing.

**Review by AlexWaygood — APPROVED** [review]
LGTM once the pytype errors are resolved!

**Comment by hauntsaninja** [comment]
Thanks! Excited for this to land

File: requirements-tests.txt
Comment (JelleZijlstra) [reply requirements-tests.txt:24]: Yes, but that only fixes the problem right now. I think long-term it's good to reflect typing-extensions changes in the stub as soon as they are merged, so that type checkers have more time to pick them up before a release.

Comment (AlexWaygood) [reply requirements-tests.txt:24]: The pytype crash is presumably because pytype has its own version of `typing.pyi` (https://github.com/google/pytype/blob/main/pytype/stubs/builtins/typing.pytd), and our stub `typing_extensions` is probably re-exporting things from `typing` that pytype doesn't have in its version of `typing`.

If this is going to happen every time a new typing_extensions feature release is cut, then I'm not sure we should be using the typing_extensions `main` branch in our CI; I feel like it would be better to pin typing_extensions and manually update it when a release is cut.

Comment (AlexWaygood) [reply requirements-tests.txt:24]: Installing the typing_extensions `main` branch in CI now causes our mypy-py37 job to crash, because the `typing_extensions` main branch uses PEP_570 syntax that can't be parsed on Python 3.7.

We have to run our mypy-py37 CI job _on_ Python 3.7 (rather than just with the `--python-version=3.7` flag), or pip will install a version of numpy in CI that's incompatible with Python 3.7, which will cause mypy to fail on e.g. our `tensorflow` stubs.

Comment (AlexWaygood) [reply requirements-tests.txt:24]: Even if we worked around the numpy issue for `mypy_test.py`, stubtest stdlib is still going to crash on py37 if we install the typing_extensions `main` branch in CI, and that's an even harder problem to solve

**Comment by rchen152** [comment]
> @rchen152 could you take a look? This is trying to implement the recent typing-extensions change where we now re-export more of typing.

I'll take a look sometime this week! Just got back from a three-week vacation, so I'm working my way through my backlog.

**Comment by rchen152** [comment]
Ok, turns out this is a pretty easy fix! pytype's copy of the typing stub was missing ForwardRef. I'll release a fixed version next week.

**Comment by rchen152** [comment]
I released pytype-2023.07.21, but it looks like typeshed can't update to it because of a dependency conflict (https://github.com/python/typeshed/actions/runs/5626188689/job/15246504135). I see that there's already a PR open in the pycln repo to fix this: https://github.com/hadialqattan/pycln/pull/209.

**Comment by AlexWaygood** [comment]
> I released pytype-2023.07.21, but it looks like typeshed can't update to it because of a dependency conflict (https://github.com/python/typeshed/actions/runs/5626188689/job/15246504135). I see that there's already a PR open in the pycln repo to fix this: [hadialqattan/pycln#209](https://github.com/hadialqattan/pycln/pull/209).

Yeah, I noticed this issue when working on #10492 yesterday :/ hopefully the pycln maintainer will be okay with  widening the libcst pin

**Comment by hauntsaninja** [comment]
(Could we just drop pycln? I think we're already using ruff + I don't like using things that pin dependencies)

File: requirements-tests.txt
Comment (AlexWaygood) [reply requirements-tests.txt:24]:
```suggestion
typing_extensions
```

**Comment by AlexWaygood** [comment]
Thanks @JelleZijlstra and @rchen152 both!

**Pull request merged**
